### ISSUE TITLE
OpenCL kernel failing to compile on Clover 

### ISSUE THREAD

**Issue by glavaux2** [arrayfire/arrayfire#2800]
I have an AMD Radeon running on Fedora 31, its device signature is:

Clover: Radeon RX 580 Series (POLARIS10, DRM 3.36.0, 5.5.8-200.fc31.x86_64, LLVM 9.0.0), 8192 MB

However running a number of examples like "surface_opencl" (master branch as of now) produces this kind of error:

ArrayFire v3.8.0 (OpenCL, 64-bit Linux, build bc37e8a9)
[0] Clover: Radeon RX 580 Series (POLARIS10, DRM 3.36.0, 5.5.8-200.fc31.x86_64, LLVM 9.0.0), 8192 MB
```
Cannot open file 1/KER17252040479444784050.cl
Cannot open file 1/KER7820264272521714025.cl
Cannot open file 1/KER5090553150507457729.cl
Cannot open file 1/KER15803858198028938426.cl
Cannot open file 1/KER6403184010839580536.cl
Cannot open file 1/KER5651956347504768133.cl
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:31:
/usr/include/clc/workitem/get_global_size.h:1:39: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_size.h:1:33: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:32:
/usr/include/clc/workitem/get_global_id.h:1:37: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_id.h:1:31: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:33:
/usr/include/clc/workitem/get_local_size.h:1:38: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_local_size.h:1:32: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:34:
/usr/include/clc/workitem/get_local_id.h:1:36: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_local_id.h:1:30: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:35:
/usr/include/clc/workitem/get_num_groups.h:1:38: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_num_groups.h:1:32: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:36:
/usr/include/clc/workitem/get_group_id.h:1:36: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_group_id.h:1:30: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:37:
/usr/include/clc/workitem/get_global_offset.h:1:41: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_offset.h:1:35: note: to match this '('
input.cl:72:26: warning: initializing '__global float *' with an expression of type 'const __global float *' discards qualifiers

 -cl-std=CL1.1 -D dim_t=long -D To=float -D Ti=float -D dim=0 
OpenCL Error (-11): Build Program Failure when calling clBuildProgram
ArrayFire Exception (Internal error:998):
OpenCL Error (-11): Build Program Failure when calling clBuildProgram
In function void af::Window::surface(const af::array&, const af::array&, const af::array&, const char*)
In file src/api/cpp/graphics.cpp:126
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
OpenCL Error (-11): Build Program Failure when calling clBuildProgram
In function void af::Window::surface(const af::array&, const af::array&, const af::array&, const char*)
In file src/api/cpp/graphics.cpp:126
Aborted (core dumped)
```

This seems to be related to this compiler argument "-D dim=0" and that lne in a OpenCL include file for example:
_CLC_DECL size_t get_global_size(uint dim);

It is clear that it should yield a compile error. Maybe ArrayFire should use better mangled macro name  like "af_dim_opencl" or something else... ?

Best regards.

**Comment by umar456** [comment]
Hey @glavaux2,

This looks like its an error in the OpenCL headers and not ArrayFire source. Looks like you are trying to print kernels in the example above. Try passing AF_JIT_KERNEL_TRACE=stdout to print the kernels. This will print it to screen. You can also pass a path to that variable and it will dump the kernels to a folder AF_JIT_KERNEL_TRACE=/tmp. This may give us more information.

**Comment by glavaux2** [comment]
Hi @umar456,

Here is the last kernel printed before the error. There is nothing wrong that I can see.

```
__kernel void
KER5651956347504768133(
float scalar0, 
__global float *in1, dim_t iInfo1_offset, 
float scalar2, 
float scalar4, 
float scalar7, 
__global float *in8, dim_t iInfo8_offset, 
float scalar9, 
float scalar11, 
float scalar17, 
float scalar22, 
float scalar27, 
__global float *out28, 
KParam oInfo, uint groups_0, uint groups_1, uint num_odims)
{


        uint groupId  = get_group_id(1) * get_num_groups(0) + get_group_id(0);
        uint threadId = get_local_id(0);
        int idx = groupId * get_local_size(0) * get_local_size(1) + threadId;
        if (idx >= oInfo.dims[3] * oInfo.strides[3]) return;
        int idx1 = idx + iInfo1_offset;
int idx8 = idx + iInfo8_offset;
float val0 = scalar0;
float val1 = in1[idx1];
float val2 = scalar2;
float val3 = __div(val1, val2);
float val4 = scalar4;
float val5 = __sub(val3, val4);
float val6 = __mul(val0, val5);
float val7 = scalar7;
float val8 = in8[idx8];
float val9 = scalar9;
float val10 = __div(val8, val9);
float val11 = scalar11;
float val12 = __sub(val10, val11);
float val13 = fabs(val12);
float val14 = __sub(val7, val13);
float val15 = __mul(val6, val14);
float val16 = __mul(val5, val5);
float val17 = scalar17;
float val18 = __add(val12, val17);
float val19 = __mul(val16, val18);
float val20 = cos(val19);
float val21 = __mul(val15, val20);
float val22 = scalar22;
float val23 = __add(val5, val22);
float val24 = __mul(val12, val23);
float val25 = sin(val24);
float val26 = __add(val21, val25);
float val27 = scalar27;
float val28 = __sub(val26, val27);
out28[idx] = val28;


}
```

The error printed by thanks to AF_OPENCL_SHOW_BUILD_INFO=1 is the following:
```
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:31:
/usr/include/clc/workitem/get_global_size.h:1:39: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_size.h:1:33: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:32:
/usr/include/clc/workitem/get_global_id.h:1:37: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_id.h:1:31: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:33:
/usr/include/clc/workitem/get_local_size.h:1:38: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_local_size.h:1:32: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:34:
/usr/include/clc/workitem/get_local_id.h:1:36: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_local_id.h:1:30: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:35:
/usr/include/clc/workitem/get_num_groups.h:1:38: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_num_groups.h:1:32: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:36:
/usr/include/clc/workitem/get_group_id.h:1:36: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_group_id.h:1:30: note: to match this '('
In file included from <built-in>:1:
In file included from /usr/include/clc/clc.h:37:
/usr/include/clc/workitem/get_global_offset.h:1:41: error: expected ')'
<command line>:4:13: note: expanded from here
/usr/include/clc/workitem/get_global_offset.h:1:35: note: to match this '('
input.cl:72:26: warning: initializing '__global float *' with an expression of type 'const __global float *' discards qualifiers

 -cl-std=CL1.1 -D dim_t=long -D To=float -D Ti=float -D dim=0 
ArrayFire Exception (Internal error:998):
OpenCL Error (-11): Build Program Failure when calling clBuildProgram
In function void af::Window::surface(const af::array&, const af::array&, const af::array&, const char*)
In file src/api/cpp/graphics.cpp:126
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
OpenCL Error (-11): Build Program Failure when calling clBuildProgram
In function void af::Window::surface(const af::array&, const af::array&, const af::array&, const char*)
In file src/api/cpp/graphics.cpp:126
Aborted (core dumped)
```

I was quoting that line in in the first report:
`_CLC_DECL size_t get_global_offset(uint dim);`

which, with `-D dim=0`, would become:
`_CLC_DECL size_t get_global_offset(uint 0);`

which is wrong.

Of course the problem is not really ArrayFire but some bad interaction with the preprocessor macro specification. I do not know the detail of OpenCL specifications w.r.t to allowed names in preprocessor macro names.

**Comment by glavaux2** [comment]
I have made a fix that solve my problem by renaming "dim" to _af_dim_ in kernels. I am not sure that I caught all instances yet.

You can find the relevant commit there at the moment: https://github.com/arrayfire/arrayfire/commit/a6d915185070bf9f69bf2f84a13b77d44e928695

**Comment by 9prady9** [comment]
Good catch, it is what it is I guess. If AMD OpenCL implementation doesn't work without this, then we would have to work around it like you are doing. However, it is very bad implementation on their part, `dim` is not even a keyword or reserved-word. This shouldn't have happened. Can you please send in a PR. I will check for other instances of where this is happening.

Thank you!

**Comment by 9prady9** [comment]
Closing as it has been addressed in relevant locations in #2802
